PCH and RelWithDebInfo

Przemog1 opened the pull request.

Przemog1 committed:
added file with quantized normals

Przemog1 committed:
Merge branch 'shader_pipeline' of https://github.com/buildaworldnet/IrrlichtBAW into shader_pipeline

Przemog1 committed:
Merge branch 'shader_pipeline' of https://github.com/buildaworldnet/IrrlichtBAW into shader_pipeline

Crisspl committed:
partial VT example update regarding divergent texture access

(have to pull matt's changes)

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into mitsuba

Crisspl committed:
VT is practically finished

Crisspl committed:
updated spirv_cross

Crisspl committed:
spirv-cross runs fine with nonuniformEXT on OpenGL now, so getting rid of workaround

Crisspl committed:
Movde VT code to builtin

Crisspl committed:
tmp

Przemog1 committed:
Merge branch 'shader_pipeline' of https://github.com/buildaworldnet/IrrlichtBAW into shader_pipeline

Crisspl committed:
Merge branch 'vao-fix' into mitsuba

Crisspl committed:
Merge branch 'vao-fix' into mitsuba

Crisspl committed:
Merge branch 'vao-fix' into mitsuba

Crisspl committed:
Turned VT functions into builtin

Crisspl committed:
Merge branch 'vao-fix' into mitsuba

Przemog1 committed:
pch

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,20 +138,24 @@ macro(irr_adjust_flags)
 		string(REPLACE "/MDd" "/MTd" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
 		string(REPLACE "/W3" "/W0" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
-		set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /Zc:__cplusplus /MP /fp:fast")
+		set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /Zc:__cplusplus /MP /fp:fast" ) #not sure if /Yu and /Yc are needed, when target_precompile_headers are used
```
probably not

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,20 +138,24 @@ macro(irr_adjust_flags)
 		string(REPLACE "/GS" "" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
 		string(REPLACE "/O2" "/Ox" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
 		string(REPLACE "/MD" "/MT" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
 		set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /GL /MP /Gy- /Zc:wchar_t /Zi /sdl- /GF /Ot /GS- /Oy /Oi /fp:fast")
```
you want the same treatment on C and CXX flags (except for __cplusplus macro)

devshgraphicsprogramming replied:
like making the O2 replace disappear

devshgraphicsprogramming replied:
`/GF` and `/Gy` are provided by `O2` so throw them out

devshgraphicsprogramming replied:
remove the `/Gy-` flag

devshgraphicsprogramming replied:
`/Ot /Oy /Oi` are provided by Ox and O2, I don't know why they were there

devshgraphicsprogramming replied:
and you don't want any `/Z` (`/Zi`) option on a release build since its pointless

devshgraphicsprogramming replied:
still

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,20 +138,24 @@ macro(irr_adjust_flags)
+		
+		# relWithDebInfo
+		#string(REPLACE "/O2" "/Ox" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
+		set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /GL /Zc:__cplusplus /Zi /fp:fast")
```
you need to handle C (non CXX) as well

devshgraphicsprogramming replied:
I don't want `/GL` on relwithdebug info, I want incremental linking instead (and incremental with minimal rebuild on debug

> Release : /GL
> Relwithdebinfo : /LCTG:INCREMENTAL
> Debug: /INCREMENTAL

A similar warning level to debug is wanted

Same `wchar_t` option for all targets

devshgraphicsprogramming replied:
`/MP /sdl-` is neede here as well

devshgraphicsprogramming replied:
finally the target also needs the `REPLACE "/MDd" "/MTd"` on both C and CXX falgs

devshgraphicsprogramming replied:
ALSO add `/Oy-` AFTER `/O2` (O2 already in the flags, so append the options)

Crisspl replied:
/GL and /LTCG is basically same thing but former is compiler flag and latter is linker flag. They must both be specified though for whole program optimization to be enabled. Otherwise im getting
```
/LTCG specified but no code generation required; remove /LTCG from the link command line to improve linker performance
```

Crisspl replied:
actually.. why dont we use ltcg:incremental in release? I guess `/GL` forces `/LTCG` and thats why it doesnt raise any warnings, but it doesnt enable `/LTCG:incremental`

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,20 +138,24 @@ macro(irr_adjust_flags)
+		set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /Zc:__cplusplus /MP /fp:fast" ) #not sure if /Yu and /Yc are needed, when target_precompile_headers are used
 		string(REPLACE "/MDd" "/MTd" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
 		string(REPLACE "/W3" "/W0" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
 		set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} /MP /fp:fast")
```
replace `/Zi` with `/ZI` maybe we'll get to use "Edit and Continue"

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -469,32 +469,37 @@ target_compile_definitions(IrrlichtServer PRIVATE
+#    else()
+#        ADD_PRECOMPILED_HEADER(IrrlichtServer pch.h FORCEINCLUDE)
+#    endif()
+#endif()
```
just kill the code

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -417,17 +417,17 @@ target_compile_definitions(Irrlicht
+#    else()
+#        ADD_PRECOMPILED_HEADER(Irrlicht pch.h FORCEINCLUDE)
+#    endif()
+#endif()
```
just kill the code

Przemog1 committed:
RelWithDebInfo for msvc

Crisspl committed:
ICPUVirtualTexture mostly done

still lives in ItexturePacker.h tho because it's easier coding this ways
Shouldnt be harder to review because of it

Crisspl committed:
a few forgotten lines

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,20 +138,31 @@ macro(irr_adjust_flags)
+		string(REPLACE "/GS" "" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
+		#string(REPLACE "/O2" "/Ox" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
+		string(REPLACE "/MDd" "/MTd" CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
+		set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} /LCTG:INCREMENTAL /MP /Gy- /Zc:wchar_t /Zi /sdl- /GF /Ot /GS- /Oy- /Oi /fp:fast")
```
too manny flags don't want `/GF /Ot /GS- /Oi`

btw. check that you can actually singlestep (or at least somewhat step) code of examples  in RelWithDebInfo builds

Crisspl committed:
ICPUVirtualTexture finished

time for GPU counterpart and utils

Crisspl committed:
very basic IGPUVirtualTexture

Crisspl committed:
Changed cpu2gpu API for smart_refctd_ptr<> to a little more reasonable

Crisspl committed:
Slight edit to FullScreenTriangle ext due to cpu2gpu API change

Crisspl committed:
fixed error when trying to create smart_refctd_dynamic_array with filling constructor

Crisspl committed:
Fixed a bug in GeneralPurposeAddrAllocatorBase move assignment op

Crisspl committed:
almost working CPU and GPU IVirtualTexture

Crisspl committed:
IVirtualTexture working

Crisspl committed:
New GLSL builtins for VT

Crisspl committed:
forgot about this

Crisspl committed:
Added bunch of utilities to IVirtualTexture

Crisspl committed:
Finished VT classes

Crisspl committed:
Overridable texture sample functions for MTL shader

Crisspl committed:
even better MTL overrides

Introducing overridable shader prologues

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into mitsuba

Crisspl committed:
pushed ex41 a little

Crisspl committed:
Fixed a few things in IVirtualTexture according to review

Przemog1 committed:
help

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,30 +138,29 @@ macro(irr_adjust_flags)
 		#string(REPLACE "/O2" "/Ox" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
 		string(REPLACE "/MD" "/MT" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
-		set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /GL /MP /Gy- /Zc:wchar_t /Zi /sdl- /GF /Ot /GS- /Oy /Oi /fp:fast")
+		set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /GL /MP /Gy- /Zc:wchar_t /sdl- /GF /Ot /GS- /Oy /Oi /fp:fast")
```
`/Ot /Oy /Oi` are provided by O2 no need for them here

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -138,30 +138,29 @@ macro(irr_adjust_flags)
 		string(REPLACE "/MDd" "/MTd" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
 		string(REPLACE "/GS" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
-		set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /LCTG:INCREMENTAL /Zc:__cplusplus /MP /Gy- /Zc:wchar_t /Zi /sdl- /GF /Ot /GS- /Oy- /Oi /fp:fast")
+		set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /LTCG:INCREMENTAL /Zc:__cplusplus,wchar_t /MP /Gy- /Zi /sdl- /Oy- /fp:fast")
```
you sure the syntax `/Zc:__cplusplus,wchar_t` works?

Przemog1 replied:
nah

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -467,10 +469,12 @@ if(IRR_PCH)
 endif()
 	
 set_target_properties(Irrlicht IrrlichtServer PROPERTIES DEBUG_POSTFIX _debug)
+set_target_properties(Irrlicht IrrlichtServer PROPERTIES RELWITHDEBDINFO_POSTFIX _relwithdebinfo)
```
I want you to get rid of the `IrrlichtServer` target completely

Crisspl committed:
Most of review comments resolved

Most of today's are probably untouched though

Crisspl committed:
More changes/fixes regarding review

Przemog1 committed:
deleted IrrlichtServer

Przemog1 committed:
Deleted IrrlichtServer - correction

Przemog1 committed:
MSVC flags fix

Crisspl committed:
New IVirtualTexture API

Crisspl committed:
reading GL exts from envvar if exists

Crisspl committed:
mitsuba loader: Creating traversal without normal reusing

Crisspl committed:
Getting rid of IVirtualTexture::E_SAMPLER_TYPE

Przemog1 committed:
MT for RelWithDebInfo

Crisspl committed:
IVirtualTexture adjustments

Crisspl committed:
GeneralpurposeAddressAllocator copy ctor

Przemog1 committed:
Foreach loop for targets

Crisspl committed:
Copy ctors for most of addr allocators

devshgraphicsprogramming commented on cmake/common.cmake:
```diff
@@ -35,7 +35,11 @@ macro(irr_create_executable_project _EXTRA_SOURCES _EXTRA_OPTIONS _EXTRA_INCLUDE
+	add_executable(${EXECUTABLE_NAME} main.cpp ${_EXTRA_SOURCES})
+	
+	set_property(TARGET ${EXECUTABLE_NAME} PROPERTY
+             MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```
what around release?

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -353,6 +353,76 @@ add_library(Irrlicht STATIC
+             MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+			 
+set_property(TARGET IlmThread PROPERTY
+             MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```
what around release?

Przemog1 replied:
Release and Relwithdebinfo are set to MT

devshgraphicsprogramming replied:
ah ok I see

Crisspl committed:
A few minor fixes/adjustments

Crisspl committed:
silly fix

Crisspl committed:
some mitsuba loader work

Przemog1 committed:
Done with MSVC

Przemog1 committed:
Removed rwdi postfix from zlib static library

devshgraphicsprogramming commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -380,8 +379,11 @@ set(IRR_TARGETS Irrlicht
-		
-			#TODO: targets must be appended to the list depending on build options
+				
+if(ENABLE_HLSL)
+	list(APPEND IRR_TARGETS HLSL)
+endif()
```
we will never support HLSL

Przemog1 replied:
it still needs to be there

Crisspl committed:
Resolved a few issues

Crisspl committed:
Got rid of unnecessary fill filter call

Crisspl committed:
merge with devsh/shader_pipeline

Crisspl committed:
merge with przemog/shader_pipeline

Crisspl committed:
fixed issues with cmake

Przemog1 committed:
Merge pull request #1 from Crisspl/cmake

Cmake
